Cross branch migration testing workflow

noahmoss opened the pull request:
Adds a new github workflow that is triggered when any migration file is updated, which does the following:
1. Checks out master
2. Starts up Metabase with a specific DB type and version as the app DB (postgres/mysql/mariadb/h2)
	* I'm reusing all the docker images that we use in driver tests here 
4. Checks out the pull request branch
5. Starts up Metabase again with the same app DB and checks that migrations run successfully

This should ensure that an instance that has previously run all the commits on master will be able to run the migration without error, and avoid future issues like https://github.com/metabase/metabase/issues/45967

Reviewers: please review carefully because I am not super well-versed with github actions. It seems to be working but I mainly pieced this together from reading the existing code so let me know if I'm missing anything.

noahmoss committed:
WIP cross branch migration testing workflow

noahmoss committed:
break a migration as a test

noahmoss committed:
undo breaking migration change

noahmoss committed:
a little cleanup

noahmoss committed:
extract core workflow as an action

noahmoss committed:
test - break a migration in a different way

noahmoss committed:
add actions/checkout@v4

noahmoss committed:
add shell:bash

noahmoss committed:
add other app dbs

noahmoss committed:
add missing steps

noahmoss committed:
fix migration

noahmoss committed:
add no-op migration that should pass

noahmoss committed:
make sure we kill metabase teh second time

noahmoss committed:
edit a mysql-only migration and make sure it's caught by the appropriate tests

noahmoss committed:
add timeouts for health checks

noahmoss committed:
remove timeouts from action since they're unsupported and lower the timeouts on the top-level jobs

noahmoss commented on .github/workflows/cross-branch-migrations.yml:
```diff
@@ -0,0 +1,144 @@
+
+  migrations-postgres-12-ee:
+    runs-on: ubuntu-22.04
+    timeout-minutes: 3
```
Apparently I can't set `timeout-minutes` on the health check steps in the action due to a GitHub limitation so I've set it here instead. It would be nicer if we could kill it automatically when the health check passes but I'm not sure how to do that and don't see any existing patterns like that

noahmoss committed:
bump up timeout to 5 mins

noahmoss commented:
Here's an example of a failed workflow in which I edited a MySQL/MariaDB-only migration in-place: https://github.com/metabase/metabase/actions/runs/10063595332/job/27818680269?pr=45998

The health check correctly times out for those DBs but passes for postgres/h2.

It's not ideal that it just times out without showing the error that Liquibase throws when Metabase starts up. If anyone has a pointer for how to adjust that I'd love to fix it.

noahmoss committed:
revert migration change

noahmoss commented on .github/workflows/migrations.yml:
```diff
@@ -1,4 +1,4 @@
-name: Migrations
+name: Lint migrations
```
Renamed this workflow to distinguish it better from the new one

iethree commented on .github/actions/cross-branch-migration-test/action.yml:
```diff
@@ -0,0 +1,43 @@
+runs:
+  using: "composite"
+  steps:
+    - name: Check out master request branch
+      uses: actions/checkout@v4
+      with:
+        ref: master
```
how about you introduce an input variable for the `base` ref so we can use this on release branches as well. It would also allow us to test migrations across release branches. 🤩

noahmoss replied:
good idea!

iethree commented on .github/actions/cross-branch-migration-test/action.yml:
```diff
@@ -0,0 +1,43 @@
+      shell: bash
+
+    - name: Wait for Metabase to start
+      run: while ! curl 'http://localhost:3000/api/health' | grep '{"status":"ok"}'; do sleep 1; done
```
I think you can export the PID from the last step, and then you can check if it is still running in this loop to be able to exit the action early if it already exited.

see https://stackoverflow.com/questions/36882936/how-to-exit-the-bash-script-when-one-of-its-background-processes-has-error

noahmoss replied:
cool — we have this pattern of polling the health check endpoint with a timeout in a few other places, so if this works I can change those spots too

iethree reviewed (changes requested):
This is fantastic 😍 , two suggestions:
- to make this reusable by other branches
- to be able to exit earlier

noahmoss committed:
use github.base_ref as a variable in the action, and break a mysql test again to validate

noahmoss committed:
try to detect whether metabase has shut down early during health checks

noahmoss committed:
tweak error message + try using nohup?

noahmoss committed:
test another approach to fix logging

noahmoss committed:
try writing mb logs to a file then explicitly tailing the file

noahmoss committed:
try explicitly printing out the logs as a separate step

noahmoss committed:
move logging printout to before we kill metabase

noahmoss committed:
try to fix mysql connection

noahmoss committed:
revert the breaking migrations change

noahmoss commented:
@iethree I think I've addressed your comments, and also fixed the logging in a roundabout way (writing the logs from the background process to a file and then `cat`ing it if we detect that Metabase exited early).

Here's a new example workflow run: https://github.com/metabase/metabase/actions/runs/10063595332/job/27818680269?pr=45998

iethree reviewed (approved):
🤩 Awesome. this fails much faster now ❤️  and is much more broadly useful

(found a couple outdated strings to update)

noahmoss committed:
Update .github/actions/cross-branch-migration-test/action.yml

Co-authored-by: Ryan Laurie <30528226+iethree@users.noreply.github.com>

noahmoss committed:
Update .github/actions/cross-branch-migration-test/action.yml

Co-authored-by: Ryan Laurie <30528226+iethree@users.noreply.github.com>

noahmoss committed:
Update .github/actions/cross-branch-migration-test/action.yml

Co-authored-by: Ryan Laurie <30528226+iethree@users.noreply.github.com>
